tesseract_command_language/include/tesseract_command_language/core/instruction.h

```diff
@@ -259,25 +282,51 @@ class Instruction
     return (*this);
   }
 
-  int getType() const { return instruction_->getType(); }
+  std::type_index getType() const
+  {
+    if (instruction_ == nullptr)
+      return std::type_index(typeid(nullptr));
+
+    return instruction_->getType();
+  }
 
   const std::string& getDescription() const { return instruction_->getDescription(); }
 
   void setDescription(const std::string& description) { instruction_->setDescription(description); }
 
   void print(const std::string& prefix = "") const { instruction_->print(prefix); }
 
-  tinyxml2::XMLElement* toXML(tinyxml2::XMLDocument& doc) const { return instruction_->toXML(doc); }
+  bool operator==(const Instruction& rhs) const
+  {
+    if (instruction_ == nullptr && rhs.instruction_ == nullptr)
+      return true;
+
+    if (instruction_ == nullptr)
+      return false;
+
+    if (rhs.instruction_ == nullptr)
+      return false;
+
+    return instruction_->operator==(*rhs.instruction_);
+  }
+
+  bool operator!=(const Instruction& rhs) const { return !operator==(rhs); }
 
   template <typename T>
   T* cast()
```

Comment:
Not sure if you wanted to address it in this PR, but we had discussed renaming the casting function to something like `as` or at least naming the two different versions the same thing. I think you already did this in a PR for the `any` type in `tesseract_common`, so I thought you might want to do the same here and for `Waypoint`

Reply:
It has been switched over to the new method `as` which returns a reference now.